zerver/lib/actions.py

```diff
@@ -3508,3 +3509,29 @@ def do_update_user_custom_profile_data(user_profile, data):
             update_or_create(user_profile=user_profile,
                              field_id=field['id'],
                              defaults={'value': field['value']})
+
+def get_unread_message_ids_per_recipient(user_profile):
+    # type: (UserProfile) -> List[Tuple[Union[Text, List[Dict[str, Any]]], List[int]]]
+    user_msgs = UserMessage.objects.filter(
+        user_profile=user_profile,
+        flags=~UserMessage.flags.read,
+    ).order_by(
+        'message__recipient__id',
+    ).values(
+        'message__id',
+        'message__recipient__id',
+        'message__recipient__type',
+        'message__recipient__type_id',
```

Comment:
I think this is the only significant problem with this PR; we want to organize by "thread", i.e. (stream, topic) pairs or for PMs, recipient_user_id lists (we probably don't need this data set to actually send the emails and other data, since we're not sending the full message bodies anyway; this whole feature is just to render unread counts and lists of threads correctly).